jgrapht-core/src/test/java/org/jgrapht/alg/tour/WarnsdorffRuleKnightTourHeuristicTest.java

```diff
@@ -0,0 +1,303 @@
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+
+package org.jgrapht.alg.tour;
+
+import org.jgrapht.alg.util.Pair;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class WarnsdorffRuleKnightTourHeuristicTest {
+
+    private KnightTour container;
+    private WarnsdorffRuleKnightTourHeuristic solver;
+
+    @Before
+    public void setup() {
+        container = new KnightTour();
+    }
+
+    private boolean checkClosed(Pair<Integer, Integer> start, Pair<Integer, Integer> end) {
+        return Math.abs(start.getFirst() - end.getFirst()) == 1 && Math.abs(start.getSecond() - end.getSecond()) == 2
+                || Math.abs(start.getFirst() - end.getFirst()) == 2 && Math.abs(start.getSecond() - end.getSecond()) == 1;
+    }
+
+    private boolean checkOpen(Pair<Integer, Integer> start, Pair<Integer, Integer> end) {
+        return !checkClosed(start, end);
+    }
+
+    private boolean checkMoveCorrectness(List<Pair<Integer, Integer>> tour, int n, int m, int shiftX, int shiftY) {
+        boolean[][] used = new boolean[shiftX + n][shiftY + m];
+        used[tour.get(0).getFirst()][tour.get(0).getSecond()] = true;
+
+        for (int i = 1; i < tour.size(); i++) {
+            if (!((Math.abs(tour.get(i - 1).getFirst() - tour.get(i).getFirst()) == 1 && Math.abs(tour.get(i - 1).getSecond() - tour.get(i).getSecond()) == 2)
+                ||
+                Math.abs(tour.get(i - 1).getFirst() - tour.get(i).getFirst()) == 2 && Math.abs(tour.get(i - 1).getSecond() - tour.get(i).getSecond()) == 1)) {
+                return false;
+            }
+
+            assertTrue(tour.get(i).getFirst() >= shiftX);
+            assertTrue(tour.get(i).getFirst() < shiftX + n);
+            assertTrue(tour.get(i).getSecond() >= shiftY);
+            assertTrue(tour.get(i).getSecond() < shiftY + m);
+
+            if (used[tour.get(i).getFirst()][tour.get(i).getSecond()]) {
+                return false;
+            }
+
+            used[tour.get(i).getFirst()][tour.get(i).getSecond()] = true;
+        }
+
+        return true;
+    }
+
+    private boolean checkStructured(List<Pair<Integer, Integer>> tour, boolean structured, int n, int m, int shiftX, int shiftY) {
+        HashSet<Pair<Pair<Integer, Integer>, Pair<Integer, Integer>>> moves = new HashSet<>();
+
+        for(int i = 1; i < tour.size(); i++) {
+            moves.add(new Pair<>(tour.get(i - 1), tour.get(i)));
+        }
+
+        return !structured || (
+            (moves.contains(new Pair<>(new Pair<>(1 + shiftX, shiftY), new Pair<>(shiftX, 2 + shiftY)))
```

Comment:
This condition is quite long and includes very similar operations. Do you think that it would be shorter, clearer and easier to read to have like an array with Δx and Δy and a for loop that checks if all 'ands' are true one at a time?